File: examples/nodejs/koskidex_client.js
/* eslint-env node */
const axios = require("axios");

/**
 * KoskidexClient - Node.js client for the Koskidex search engine.
 */
class KoskidexClient {
  /**
   * @param {string} host - The base URL of the Koskidex server.
   * @param {string} apiKey - Optional API key for authentication.
   */
  constructor(host = "http://localhost:7700", apiKey = "") {
    this.client = axios.create({
      baseURL: host,
      headers: apiKey ? { Authorization: `Bearer ${apiKey}` } : {},
    });
  }

  /**
   * @param {string} name - The name of the index to create.
   * @returns {Promise<any>}
   */
  async createIndex(name) {
    const res = await this.client.post("/indexes", { name });
    return res.data;
  }

  /**
   * @param {string} indexName - The name of the index.
   * @param {Array<Object>} documents - Array of documents to add.
   * @returns {Promise<any>}
   */
  async addDocuments(indexName, documents) {
    const res = await this.client.post(
      `/indexes/${indexName}/documents`,
      documents,
    );
    return res.data;
  }

  /**
   * @param {string} indexName - The name of the index.
   * @param {Object} document - A single document object to add.
   * @returns {Promise<any>}
   */
  async addDocument(indexName, document) {
    const res = await this.client.post(
      `/indexes/${indexName}/documents`,
      document,
    );
    return res.data;
  }

  /**
   * @param {string} indexName - The name of the index.
   * @param {string} query - The search query string.
   * @returns {Promise<any>}
   */
  async search(indexName, query) {
    const res = await this.client.get(`/indexes/${indexName}/search`, {
      params: { q: query },
    });
    return res.data;
  }
}

// ============== USAGE EXAMPLE ==============
async function runExample() {
  const koski = new KoskidexClient("http://localhost:7700");

  try {
    console.log("Creating 'products' index...");
    await koski.createIndex("products");
  } catch (/** @type {any} */ e) {
    console.log(
      "Index already exists or error:",
      e.response?.data || e.message,
    );
  }

  console.log("\nAdding documents in batch...");
  await koski.addDocuments("products", [
    { id: "100", name: "MacBook Pro 14", price: 1999, category: "Laptops" },
    { id: "101", name: "iPhone 15 Pro", price: 999, category: "Smartphones" },
    { id: "102", name: "iPad Air", price: 599, category: "Tablets" },
  ]);

  console.log("\nSearching: 'macbok' (Typo test)...");
  const results = await koski.search("products", "macbok");

  console.log("\nResults found:");
  console.log(JSON.stringify(results, null, 2));
}

if (require.main === module) {
  runExample().catch((/** @type {any} */ err) => {
    console.error("Example failed:", err.message);
  });
}

module.exports = KoskidexClient;

File: examples/python/koskidex_client.py
import requests
import json

class KoskidexClient:
    def __init__(self, host='http://localhost:7700', api_key=None):
        self.host = host.rstrip('/')
        self.session = requests.Session()
        if api_key:
            self.session.headers.update({'Authorization': f'Bearer {api_key}'})

    def create_index(self, name: str):
        res = self.session.post(f"{self.host}/indexes", json={"name": name})
        res.raise_for_status()
        return res.json()

    def add_documents(self, index_name: str, documents: list):
        res = self.session.post(f"{self.host}/indexes/{index_name}/documents", json=documents)
        res.raise_for_status()
        return res.json()

    def add_document(self, index_name: str, document: dict):
        res = self.session.post(f"{self.host}/indexes/{index_name}/documents", json=document)
        res.raise_for_status()
        return res.json()

    def search(self, index_name: str, query: str):
        res = self.session.get(f"{self.host}/indexes/{index_name}/search", params={"q": query})
        res.raise_for_status()
        return res.json()

# ============== USAGE EXAMPLE ==============
if __name__ == "__main__":
    client = KoskidexClient()

    try:
        print("Creating 'articles' index...")
        client.create_index("articles")
    except Exception as e:
        print("Index might already exist.")

    print("Adding documents in batch...")
    client.add_documents("articles", [
        {"id": "a1", "title": "Understanding Python Decorators", "tags": "python, functions"},
        {"id": "a2", "title": "A Guide to Go Interfaces", "tags": "golang, interfaces"},
        {"id": "a3", "title": "Building REST APIs", "tags": "api, rest, web"}
    ])

    print("Searching: 'gude' (Typo test)...")
    results = client.search("articles", "gude")
    
    print("\nResults found:")
    print(json.dumps(results, indent=2))
